python/ray/data/_internal/batcher.py

```diff
@@ -73,6 +74,19 @@ def add(self, block: Block):
             block: Block to add to the block buffer.
         """
         if BlockAccessor.for_block(block).num_rows() > 0:
+            import pyarrow
+
+            # pyarrow.Table.slice is slow when the table has many chunks
+            # so we combine chunks into a single one to make slice faster
+            # with the cost of an extra copy.
+            # See https://github.com/ray-project/ray/issues/31108 for more details.
+            if (
+                isinstance(block, pyarrow.Table)
+                and block.num_columns > 0
+                and block.column(0).num_chunks > 1
+            ):
+                block = transform_pyarrow.combine_chunks(block)
```

Comment:
can we structure the code to only do `combine_chunks` when necessary? 

* In `next_batch()` below, we call `slice()` to get one batch from block. Can we call `combine_chunks` there instead? We don't need to call `combine_chunks` if the block is not sliced.

* How many `num_chunks` we saw in benchmark? Can we have a minimal threshold? I feel `num_chunks > 1` is a bit too aggressive.

Reply:
In the benchmark we have 8k+ chunks. But with my test, even with 10 big chunks, combine_chunks first is still faster.

Reply:
what's the size of big chunks? Shall we decide to combine chunks based on
* number of chunks
* size of each chunk

Reply:
Discussed offline with @jjyao, we decided to go with current approach, and expose a constant `MIN_NUM_CHUNKS_TO_TRIGGER_COMBINE_CHUNKS`, so we can change the constant easily when debugging the issue later.